### PR TITLE
Migrate to TypeScript

### PR DESCRIPTION
**Pull request by herberttn**
#### Decorators
- The `@OdinConfig` decorator is now `@Configuration`.
- The `@PostConstruct` decorator is now `@Initializer`.
- The `@Singleton` decorator is now `@Injectable({ singleton: true })`.

#### Fallbacks
- `Bundle.get` now returns `undefined` instead of `null`.
- `Container.get` now returns `undefined` instead of `null`.
- `Container.provide` now returns `undefined` instead of `null`.
- `CustomProvider.resolve` now returns `undefined` instead of `null`.
- `Registry.get` now returns `undefined` instead of `null`.

#### Renames
- Renamed `InjectableDef` to `Descriptor`.
- Renamed `InjectableDef.args` to `Descriptor.options`.
- Renamed `InjectableDef.definition` to `Descriptor.injectable`.
- Renamed `Bundle.checkHas` to `Bundle.validateRegistration`.
- Renamed `Registry.checkHas` to `Registry.validateRegistration`.
- Renamed `Secrets.getPostContruct` to `Secrets.getInitializer`.
- Renamed `Secrets.setPostContruct` to `Secrets.setInitializer`.

#### Removals
- Removed `Bundle.getId`.
The user should not have to decide which identifier to use, all of them should be accepted everywhere.

- Removed `Registry.getId`.
The user should not have to decide which identifier to use, all of them should be accepted everywhere.

- Removed `Registry.hasName`.
It has been covered by the new `has` implementation.

- Removed `Secrets.getNamed`, `Secrets.isNamed` and `Secrets.setNamed`.
It was the wrong solution for a simple problem, which has now been resolved.
Feel free to implement/manage your own secrets.

- Removed `Secrets.getWrapper`, `Secrets.isWrapper` and `Secrets.setWrapper`.
The `wrapper` had no purpose within odin itself.
Feel free to implement/manage your own secrets.

- Removed `Secrets.isCustom` and `Secrets.setCustom`.
The `custom` flag had no purpose within odin itself.
Feel free to implement/manage your own secrets.

### Test coverage
image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,37 +8,42 @@
...
+    "build": "tsc --build tsconfig-refs.json",
     "clean": "rimraf --glob lib **/.eslintcache **/*.log **/*.tsbuildinfo **/coverage",
     "lint": "eslint . --ext .js,.jsx,.cjs,.mjs,.ts,.tsx,.cts,.mts --cache --max-warnings 0",
     "release": "semantic-release",
     "test": "vitest run --coverage",
     "test:dev": "vitest dev --coverage",
     "test:ui": "vitest dev --coverage --ui"
   },
+  "dependencies": {
+    "debug": "^4.3.4"
```
Comment (mfklauberg) [inline package.json:26]: pin dependency;

File: package.json [hunk 2]
```diff
@@ -8,37 +8,42 @@
...
     "@semantic-release/changelog": "6.0.3",
     "@semantic-release/git": "10.0.1",
-    "@typescript-eslint/eslint-plugin": "5.60.0",
-    "@typescript-eslint/parser": "5.60.0",
-    "@vitest/coverage-v8": "0.32.2",
-    "@vitest/ui": "0.32.2",
-    "esbuild": "0.18.7",
-    "esbuild-decorators": "1.0.0",
-    "eslint": "8.43.0",
+    "@types/debug": "^4.1.8",
```
Comment (mfklauberg) [inline package.json:31]: pin dependency;

File: src/common/decorators.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+ * @param instance the injectable instance where to invoke the eagers.
+ */
+function invokeEagers(injectable: Injectable, instance: Instance): void {
+  for (const eager of Secrets.getEagers(injectable)) {
+    logger.decorators.debug('invoking eager:', eager);
+    eat(instance[eager]);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  function eat(_: any): void {
```
Comment (mfklauberg) [inline src/common/decorators.ts:119]: imo `noop` would be more descriptive;

File: src/common/normalizers.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Normalizes the domain.
+ * In strict mode, normalizes the domain to lower case.
+ *
+ * @param domain the domain to normalize.
+ * @returns the normalized domain.
+ */
+function normalizeDomain<T>(domain: T): T {
+  // logger.debug('normalizing domain:', domain, { strict: configuration.isStrict() });
```
Comment (mfklauberg) [inline src/common/normalizers.ts:11]: uncomment or remove;

File: src/common/normalizers.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Normalizes the name or identifier.
+ * In strict mode, normalizes the name or identifier to lower case.
+ *
+ * @param nameOrIdentifier the name or identifier to normalize.
+ * @returns the normalized name or identifier.
+ */
+function normalizeNameOrIdentifier<T>(nameOrIdentifier: T): T {
+  // logger.debug('normalizing name or identifier:', name, { strict: configuration.isStrict() });
```
Comment (mfklauberg) [inline src/common/normalizers.ts:27]: uncomment or remove;

File: src/common/normalizers.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import { configuration } from '../singletons/configuration.js';
+
+/**
+ * Normalizes the domain.
+ * In strict mode, normalizes the domain to lower case.
+ *
+ * @param domain the domain to normalize.
+ * @returns the normalized domain.
+ */
+function normalizeDomain<T>(domain: T): T {
```
Comment (mfklauberg) [inline src/common/normalizers.ts:10]: Should we allow generic domains? afaik only strings are used, and I think that we should enforce it.

File: src/common/normalizers.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+/**
+ * Normalizes the name or identifier.
+ * In strict mode, normalizes the name or identifier to lower case.
+ *
+ * @param nameOrIdentifier the name or identifier to normalize.
+ * @returns the normalized name or identifier.
+ */
+function normalizeNameOrIdentifier<T>(nameOrIdentifier: T): T {
```
Comment (mfklauberg) [inline src/common/normalizers.ts:26]: Same as the comment for domains;

File: src/common/validators.ts [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+ * @param options the options to customize the validation.
+ *
+ * @throws Error if {@link decoratorOptions} is empty.
+ * @throws Error if {@link decoratorOptions} is not an object.
+ * @throws Error if {@link decoratorOptions} has unknown properties, according to {@link options.allowUnknown} and {@link options.known}.
+ * @throws Error if {@link decoratorOptions} does not have the required properties, according to {@link options.required}.
+ * @throws Error if {@link options} is not an object.
+ */
+function validateDecoratorOptions(decoratorOptions: any, options: ValidateDecoratorOptions): void {
+  // logger.debug('validating decorator options:', { decoratorOptions, options });
```
Comment (mfklauberg) [inline src/common/validators.ts:74]: uncomment or remove;

File: src/common/validators.ts [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+  if (typeof domain === 'boolean') {
+    throw new Error(logger.createMessage(`Invalid domain. It cannot be a boolean.`));
+  }
+
+  if (typeof domain === 'object') {
+    if (Array.isArray(domain)) {
+      throw new Error(logger.createMessage(`Invalid domain. It cannot be an array.`));
+    }
+    throw new Error(logger.createMessage(`Invalid domain. It cannot be an object.`));
```
Comment (mfklauberg) [inline src/common/validators.ts:159]: Wouldn't it be easier to just check for string a throw if not?

File: src/common/validators.ts [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+ * @throws Error if the {@link nameOrIdentifier} is blank.
+ * @throws Error if the {@link nameOrIdentifier} is empty.
+ * @throws Error if the {@link nameOrIdentifier} is a boolean.
+ * @throws Error if the {@link nameOrIdentifier} is an array.
+ * @throws Error if the {@link nameOrIdentifier} is an object.
+ * @throws Error if the {@link nameOrIdentifier} has chunks.
+ * @throws Error if the {@link nameOrIdentifier} has spaces.
+ */
+function validateNameOrIdentifier(nameOrIdentifier: any): void {
+  // logger.debug('validating name or identifier:', nameOrIdentifier);
```
Comment (mfklauberg) [inline src/common/validators.ts:199]: uncomment or remove;

File: src/common/validators.ts [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+ * @throws Error if the {@link domain} is a boolean.
+ * @throws Error if the {@link domain} is an array.
+ * @throws Error if the {@link domain} is an object.
+ * @throws Error if the {@link domain} has chunks and {@link options.allowHierarchy} is {@link false}.
+ * @throws Error if any of the {@link domain} chunks is blank.
+ * @throws Error if any of the {@link domain} chunks is empty.
+ * @throws Error if any of the {@link domain} chunks has spaces.
+ */
+function validateDomain(domain: any, options?: ValidateDomainOptions): void {
+  // logger.debug('validating domain:', domain);
```
Comment (mfklauberg) [inline src/common/validators.ts:143]: uncomment or remove;

File: src/common/validators.ts [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+  if (typeof nameOrIdentifier === 'boolean') {
+    throw new Error(logger.createMessage(`Invalid name or identifier. It cannot be a boolean.`));
+  }
+
+  if (typeof nameOrIdentifier === 'object') {
+    if (Array.isArray(nameOrIdentifier)) {
+      throw new Error(logger.createMessage(`Invalid name or identifier. It cannot be an array.`));
+    }
+    throw new Error(logger.createMessage(`Invalid name or identifier. It cannot be an object.`));
+  }
```
Comment (mfklauberg) [inline src/common/validators.ts:214]: Same as the domain comment, I think that we should enforce both domains and identifiers as strings.

**Review by mfklauberg — CHANGES_REQUESTED** [review]
Noticed some inconsistencies in methods and functions documented with jsdocs, as some methods document their `@throws` and others don't.

File: package.json (hunk 1 above)
Comment (herberttn) [reply package.json:26]: open source packages should not pin deps, it hinders the users ability to get security fixes

File: package.json (hunk 2 above)
Comment (herberttn) [reply package.json:31]: open source packages should not pin deps, it hinders the users ability to get security fixes

File: src/common/validators.ts (hunk 9 above)
Comment (herberttn) [reply src/common/validators.ts:159]: Where's the fun

**Review by mfklauberg — APPROVED** [review]

**Pull request merged**